config-model/src/main/java/com/yahoo/vespa/model/application/validation/change/RestartOnDeployForSidecarValidator.java

```diff
@@ -0,0 +1,120 @@
+                    var message = "Need to restart services in %s due to added sidecar '%s'"
+                            .formatted(clusterId, nextSidecar.name());
+                    addRestartAction(context, clusterId, message);
+                } else {
+                    var previousSidecar = matchingPreviousSidecar.get();
+
+                    if (!previousSidecar.equals(nextSidecar)) {
+                        var sidecarChanges = findSidecarChanges(previousSidecar, nextSidecar);
+                        var restartMessage = "Need to restart services in %s due to changed sidecar '%s' (%s)"
+                                .formatted(clusterId, nextSidecar.name(), sidecarChanges);
+                        addRestartAction(context, clusterId, restartMessage);
+                    }
+                }
+            }
+        }
+    }
+
+    private Optional<ClusterSpec> findClusterSpec(VespaModel model, ClusterSpec.Id clusterId) {
+        return model.allClusters().stream()
+                .filter(c -> c.id().equals(clusterId))
+                .findFirst();
+    }
+
+    private String findSidecarChanges(SidecarSpec from, SidecarSpec to) {
+        var changes = new ArrayList<String>();
+        var fromResources = from.resources();
+        var toResources = to.resources();
+
+        if (from.id() != to.id()) {
+            changes.add("id: " + from.id() + " -> " + to.id());
+        }
+        if (!from.name().equals(to.name())) {
+            changes.add("name: '" + from.name() + "' -> '" + to.name() + "'");
+        }
+        if (!from.image().equals(to.image())) {
+            changes.add("image: '" + from.image() + "' -> '" + to.image() + "'");
+        }
+
+        if (fromResources.maxCpu() != toResources.maxCpu()) {
+            changes.add("maxCpu: " + fromResources.maxCpu() + " -> " + toResources.maxCpu());
+        }
+        if (fromResources.minCpu() != toResources.minCpu()) {
+            changes.add("minCpu: " + fromResources.minCpu() + " -> " + toResources.minCpu());
+        }
+        if (fromResources.memoryGiB() != toResources.memoryGiB()) {
+            changes.add("memoryGiB: " + fromResources.memoryGiB() + " -> " + toResources.memoryGiB());
+        }
+        if (fromResources.hasGpu() != toResources.hasGpu()) {
+            changes.add("hasGpu: " + fromResources.hasGpu() + " -> " + toResources.hasGpu());
+        }
+
+        if (!from.volumeMounts().equals(to.volumeMounts())) {
+            changes.add("volumeMounts: " + from.volumeMounts() + " -> " + to.volumeMounts());
+        }
+        if (!from.envs().equals(to.envs())) {
+            changes.add("envs: " + from.envs() + " -> " + to.envs());
+        }
+        if (!from.command().equals(to.command())) {
+            changes.add("command: " + from.command() + " -> " + to.command());
+        }
+
+        return String.join(", ", changes);
+    }
+
+    private void addRestartAction(ChangeContext context, ClusterSpec.Id clusterId, String message) {
```

Comment:
Take a look at which ctor is used for the validators that are modified in https://github.com/vespa-engine/vespa/pull/35517.